subsys/usb/usb_c/usbc_pe_common.c

```diff
@@ -720,6 +732,216 @@ void pe_drs_send_swap_exit(void *obj)
+				pe_set_state(dev, PE_SNK_WAIT_FOR_CAPABILITIES);
+			}
+			/*
+			 * The Policy Engine Shall transition to the
+			 * PE_SRC_Hard_Reset state when:
+			 *      1: A SenderResponseTimer timeout occurs.
+			 *      2: OR the Protocol Layer indicates that a
+			 *         transmission error has occurred.
+			 *         (Handled in pe_report_error)
+			 */
+			else if (usbc_timer_expired(&pe->pd_t_sender_response)) {
+				pe_set_state(dev, PE_SNK_HARD_RESET);
+			}
+			break;
+		}
+	}
+}
+
+void pe_soft_reset_exit(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+
+	/* Stop Sender Response Timer */
+	usbc_timer_stop(&pe->pd_t_sender_response);
+}
+
+/**
+ * @brief PE_Send_Soft_Reset Entry State
+ */
+void pe_send_soft_reset_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	LOG_INF("PE_SNK_Send_Soft_Reset");
+
+	/* Reset Protocol Layer */
+	prl_reset(dev);
+	atomic_set_bit(pe->flags, PE_FLAGS_SEND_SOFT_RESET);
+}
+
+/**
+ * @brief PE_Send_Soft_Reset Run State
+ */
+void pe_send_soft_reset_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+	struct usbc_port_data *data = dev->data;
+	struct protocol_layer_rx_t *prl_rx = data->prl_rx;
+	union pd_header header;
+
+	if (prl_is_running(dev) == false) {
+		return;
+	}
+
+	if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_SEND_SOFT_RESET)) {
+		/* Send Soft Reset message */
+		pe_send_ctrl_msg(dev, pe->soft_reset_sop, PD_CTRL_SOFT_RESET);
+		return;
+	}
+
+	if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_MSG_DISCARDED)) {
+		/* Inform Device Policy Manager that the message was discarded */
+		policy_notify(dev, MSG_DISCARDED);
+		pe_set_state(dev, PE_SNK_READY);
+	} else if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_TX_COMPLETE)) {
+		/* Start SenderResponse timer */
+		usbc_timer_start(&pe->pd_t_sender_response);
+	}
+	/*
+	 * The Policy Engine Shall transition to the PE_SNK_Wait_for_Capabilities
+	 * state when:
+	 *      1: An Accept Message has been received on SOP
+	 */
+	else if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_MSG_RECEIVED)) {
+		header = prl_rx->emsg.header;
+
+		if (received_control_message(dev, header, PD_CTRL_ACCEPT)) {
+			pe_set_state(dev, PE_SNK_WAIT_FOR_CAPABILITIES);
+		}
+	}
+	/*
+	 * The Policy Engine Shall transition to the PE_SNK_Hard_Reset state when:
+	 *      1: A SenderResponseTimer timeout occurs (Handled in pe_report_error function)
+	 *      2: Or the Protocol Layer indicates that a transmission error has occurred
+	 */
+	else if (usbc_timer_expired(&pe->pd_t_sender_response)) {
+		pe_set_state(dev, PE_SNK_HARD_RESET);
+	}
+}
+
+/**
+ * @brief PE_Send_Soft_Reset Exit State
+ */
+void pe_send_soft_reset_exit(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+
+	/* Stop Sender Response Timer */
+	usbc_timer_stop(&pe->pd_t_sender_response);
```

Comment:
This can be another PR, but we'll be doing this a lot. Maybe we should have a superstate for SenderResponseTimer like we do downstream or at least a utility function.

Reply:
I agree and have been looking into adding a super state.